controllers/gateway/gateway.go

```diff
@@ -0,0 +1,177 @@
+			Status: metav1.ConditionTrue,
+			Reason: string(gateway_v1.GatewayReasonAccepted),
+		},
+		metav1.Condition{
+			Type:   string(gateway_v1.GatewayConditionProgrammed),
+			Status: metav1.ConditionTrue,
+			Reason: string(gateway_v1.GatewayReasonProgrammed),
+		},
+	)
+
+	if !equality.Semantic.DeepEqual(gateway.Status, previousStatus) {
+		if err := c.Status().Update(ctx, gateway); err != nil {
+			logger.Error(err, "couldn't update Gateway status", "name", gateway.Name)
+		}
+	}
+}
+
+// ensureListenerStatusExists ensures that the elements of g.Status.Listeners correspond to the
+// elements of g.Spec.Listeners.
+func ensureListenerStatusExists(g *gateway_v1.Gateway) {
+	// Check to see if the listeners status already matches.
+	if len(g.Status.Listeners) == len(g.Spec.Listeners) {
+		ok := true
+		for i := range len(g.Spec.Listeners) {
+			if g.Status.Listeners[i].Name != g.Spec.Listeners[i].Name {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			return
+		}
+	}
+
+	// Allocate new listeners status and copy over any existing status.
+	listenerStatusMap := make(map[string]gateway_v1.ListenerStatus)
+	for _, ls := range g.Status.Listeners {
+		listenerStatusMap[string(ls.Name)] = ls
+	}
+	g.Status.Listeners = make([]gateway_v1.ListenerStatus, len(g.Spec.Listeners))
+	for i := range len(g.Spec.Listeners) {
+		g.Status.Listeners[i] = listenerStatusMap[string(g.Spec.Listeners[i].Name)]
+	}
+}
+
+var (
+	gatewayAddressTypeIPAddress = gateway_v1.AddressType("IPAddress")
+	gatewayAddressTypeHostname  = gateway_v1.AddressType("Hostname")
+)
+
+func updateGatewayAddresses(
+	o *objects,
+	gateway *gateway_v1.Gateway,
+	serviceName types.NamespacedName,
+) {
+	// Copy the external addresses from the "pomerium-proxy" service, if it exists.
+	proxy := o.Services[serviceName]
+	if proxy == nil {
+		return
+	}
+	gateway.Status.Addresses = make([]gateway_v1.GatewayStatusAddress, 0, len(proxy.Status.LoadBalancer.Ingress))
+	for _, ingress := range proxy.Status.LoadBalancer.Ingress {
+		if ingress.IP != "" {
+			gateway.Status.Addresses = append(gateway.Status.Addresses, gateway_v1.GatewayStatusAddress{
+				Type:  &gatewayAddressTypeIPAddress,
+				Value: ingress.IP,
+			})
+		} else if ingress.Hostname != "" {
+			gateway.Status.Addresses = append(gateway.Status.Addresses, gateway_v1.GatewayStatusAddress{
+				Type:  &gatewayAddressTypeHostname,
+				Value: ingress.Hostname,
+			})
+		}
+	}
+}
+
+func upsertGatewayConditions(g *gateway_v1.Gateway, conditions ...metav1.Condition) (modified bool) {
+	return upsertConditions(&g.Status.Conditions, g.Generation, conditions...)
+}
+
+func (c *gatewayController) updateModifiedHTTPRouteStatus(
+	ctx context.Context, s []httpRouteAndOriginalStatus,
+) {
+	for _, r := range s {
+		if !equality.Semantic.DeepEqual(r.route.Status, r.originalStatus) {
+			if err := c.Status().Update(ctx, r.route); err != nil {
+				golog.Printf("couldn't update status for %q: %v", r.route.Name, err)
```

Comment:
Need to return these errors from the reconcile function, otherwise this can get stuck out of date if the reconciler isn't re-queued

Reply:
Thanks; done.